security/advancedtls/crl_provider.go

```diff
@@ -0,0 +1,236 @@
+	// Callers are expected to use the returned value as read-only.
+	CRL(cert *x509.Certificate) (*CRL, error)
+}
+
+// StaticCRLProvider implements CRLProvider interface by accepting raw content
+// of CRL files at creation time and storing parsed CRL structs  in-memory.
+type StaticCRLProvider struct {
+	crls map[string]*CRL
+}
+
+// MakeStaticCRLProvider processes raw content of CRL files, adds parsed CRL
+// structs into in-memory, and returns a new instance of the StaticCRLProvider.
+func MakeStaticCRLProvider(rawCRLs [][]byte) *StaticCRLProvider {
+	p := StaticCRLProvider{}
+	p.crls = make(map[string]*CRL)
+	for idx, rawCRL := range rawCRLs {
+		cRL, err := NewCRL(rawCRL)
+		if err != nil {
+			grpclogLogger.Warningf("Can't parse raw CRL number %v from the slice: %v", idx, err)
+			continue
+		}
+		p.addCRL(cRL)
+	}
+	return &p
+}
+
+// AddCRL adds/updates provided CRL to in-memory storage.
+func (p *StaticCRLProvider) addCRL(crl *CRL) {
+	key := crl.certList.Issuer.ToRDNSequence().String()
+	p.crls[key] = crl
+}
+
+// CRL returns CRL struct if it was previously loaded by calling AddCRL.
+func (p *StaticCRLProvider) CRL(cert *x509.Certificate) (*CRL, error) {
+	return p.crls[cert.Issuer.ToRDNSequence().String()], nil
+}
+
+// Options represents a data structure holding a
+// configuration for FileWatcherCRLProvider.
+type Options struct {
+	CRLDirectory               string          // Path of the directory containing CRL files
+	RefreshDuration            time.Duration   // Time interval between CRLDirectory scans
+	crlReloadingFailedCallback func(err error) // Custom callback executed when a CRL file can’t be processed
+}
+
+// FileWatcherCRLProvider implements the CRLProvider interface by periodically scanning
+// CRLDirectory (see Options) and storing CRL structs in-memory
+type FileWatcherCRLProvider struct {
+	crls map[string]*CRL
+	opts Options
+	mu   sync.Mutex
+	done chan struct{}
+}
+
+// MakeFileWatcherCRLProvider returns a new instance of the
+// FileWatcherCRLProvider. It uses Options to validate and apply configuration
+// required for creating a new instance.
+func MakeFileWatcherCRLProvider(o Options) (*FileWatcherCRLProvider, error) {
+	if err := o.validate(); err != nil {
+		return nil, err
+	}
+	done := make(chan struct{})
+	provider := &FileWatcherCRLProvider{
+		crls: make(map[string]*CRL),
+		opts: o,
+		done: done,
+	}
+	go provider.run()
+	return provider, nil
+}
+
+func (o *Options) validate() error {
+	// Checks relates to CRLDirectory.
+	if o.CRLDirectory == "" {
+		return fmt.Errorf("advancedtls: CRLDirectory needs to be specified")
+	}
+	fileInfo, err := os.Stat(o.CRLDirectory)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return fmt.Errorf("advancedtls: CRLDirectory %v does not exist", o.CRLDirectory)
+		}
+		return err
+	}
+	if !fileInfo.IsDir() {
+		return fmt.Errorf("advancedtls: CRLDirectory %v is not a directory", o.CRLDirectory)
+	}
+	_, err = os.Open(o.CRLDirectory)
```

Comment:
Yeah I think it makes sense. Instead of throwing a number of custom error messages now I just have one but include original error from os.Read
Yes, the validation that directory exists happens at creation time (background scan also kicks out automatically). 'Lazy' part of the scenario is - the directory can be empty and CRL files can be shipped later.